doc/languages-frameworks/javascript.section.md

```diff
@@ -275,6 +275,51 @@ mkYarnPackage rec {
   - The `echo 9` steps comes from this answer: <https://stackoverflow.com/a/49139496>
   - Exporting the headers in `npm_config_nodedir` comes from this issue: <https://github.com/nodejs/node-gyp/issues/1191#issuecomment-301243919>
 
+### buildNpmPackage {#javascript-buildNpmPackage}
+
+`buildNpmPackage` allows you to package npm-based projects in Nixpkgs without the use of an auto-generated dependencies file ([node2nix](#javascript-node2nix)):
+
+```nix
+{ lib, buildNpmPackage, fetchFromGitHub }:
+
+buildNpmPackage rec {
+  pname = "flood";
+  version = "4.7.0";
+
+  src = fetchFromGitHub {
+    owner = "jesec";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-BR+ZGkBBfd0dSQqAvujsbgsEPFYw/ThrylxUbOksYxM=";
+  };
+
+  npmDepsHash = "sha256-s8SpZY/1tKZVd3vt7sA9vsqHvEaNORQBMrSyhWpj048=";
+
+  npmBuildScript = "build";
+
+  NODE_OPTIONS = "--openssl-legacy-provider";
+
+  npmPackFlagsArray = [ "--ignore-scripts" ];
+
+  meta = with lib; {
+    description = "A modern web UI for various torrent clients with a Node.js backend and React frontend";
+    homepage = "https://flood.js.org";
+    license = licenses.gpl3Only;
+    maintainers = with maintainers; [ winter ];
+  };
+}
+```
+
+#### Arguments {#javascript-buildNpmPackage-arguments}
+
+* `npmDepsHash`: The output hash of the dependencies for this project. Can be calculated in advance with `prefetch-npm-deps`.
+* `makeCacheWritable`: Whether to make the cache writable prior to installing dependencies. Don't set this unless npm tries to write to the cache directory, as it can slow down the build.
+* `npmBuildScript`: The script to run to build the project. Defaults to `build`.
+* `npmFlagsArray`: Flags to pass to all npm commands.
+* `npmInstallFlagsArray`: Flags to pass to `npm ci`.
+* `npmBuildFlagsArray`: Flags to pass to `npm run ${npmBuildScript}`.
+* `npmPackFlagsArray`: Flags to pass to `npm pack`.
```

Comment:
> Is this encoding types in attribute names? If so, please let's not.

I did this because other similar arguments (`cmakeFlagsArray` for example) do the same.

> Annotate the types explicitly instead

How do you suggest doing that in the docs? Just say "an array of flags ..." or something? (They're already "annotated" in the code as a default parameter.)

Unfortunately, because we don't have structured attributes, users can pass both a string and an array to this argument, and it'll work all the same, so IMO the extra suffix doesn't hurt for that reason, but I see what you mean.

> We don't have a convention in the manuals yet, but I think it's good taste to use a definition list for parameters, because it represents the semantics correctly and also allows to take care of how we display it independently of contents.

(Are you saying what I'm doing is this?)

Reply:
>>    Annotate the types explicitly instead
> How do you suggest doing that in the docs? Just say "an array of flags ..." 

Somehow... we don't have a convention. `_(array of strings)_` may be one way. We probably should have examples as well, since this is not clear on its own. Annotation in the code is not enough. Hopefully no one ever has to look at the code except for changing it.

> Definition lists
> (Are you saying what I'm doing is this?)

Ah, sorry, that was unclear. I wanted to ask you to use a definition list instead of a list.

Reply:
> I wanted to ask you to use a definition list instead of a list.

What is that? Does that use headers instead of bullets for the list items or something?

Reply:
Also: how do you feel about warning if a non-array is passed? I think that would also be good for UX.

Reply:
> What is that? Does that use headers instead of bullets for the list items or something?

https://www.markdownguide.org/extended-syntax/#definition-lists

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl

> Also: how do you feel about warning if a non-array is passed? I think that would also be good for UX.

@infinisil can answer this better or refer this to someone adequately qualified. I try to focus only on documentation to get anything done at all.

Reply:
It should definitely _not_ have the `Array` suffix, so it should be just `npmPackFlags`, etc. While `configureFlagsArray` and co. do exist, those are conventionally _bash_ variables and not Nix variables, you need to run `configureFlagsArray+=( ... )` in the build script to append elements to those. `configureFlagsArray` was [originally introduced](0b7e25616262410d35a96d3bdf39a6c204467470) because `configureFlags` (the Nix variable) doesn't allow passing arguments with spaces in them, since it just gets turned into an environment variable using `toString`, which merely concatenates with `" "`'s, without handling escaping at all.

This is legacy however, we can do much better with a new function: Don't just pass these arguments to the `mkDerivation` function as-is, instead escape them properly with `npmPackFlags = lib.escapeShellArgs npmPackFlags`, then read it into a bash array from the builder using `eval "npmPackFlags=($npmPackFlags)"`. Doing this will propagate strings from a Nix list to a bash array as-is, without any escaping issues. See [this PR](https://github.com/NixOS/nixpkgs/pull/164662) for an example.